cluster-autoscaler/estimator/cluster_capacity_threshold_test.go

```diff
@@ -0,0 +1,62 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package estimator
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestNewClusterCapacityThreshold(t *testing.T) {
+	tests := []struct {
+		name                string
+		wantThreshold       int
+		contextMaxNodes     int
+		contextCurrentNodes int
+	}{
+		{
+			name:                "computes available capacity",
+			wantThreshold:       5,
+			contextMaxNodes:     10,
+			contextCurrentNodes: 5,
+		},
+		{
+			name:                "for not defined cluster max nodes returns 0",
+			wantThreshold:       0,
+			contextMaxNodes:     0,
+			contextCurrentNodes: 10,
+		},
+		{
+			name:                "for negative total capacity returns 0",
```

Comment:
I think in this case the 0 should mean "You cannot add any nodes at all". This is a bit problematic with the current definition of 0. Maybe we can change the semantic so that "You can add however many nodes you want" would be represented by an arbitrary large number, eg. 10^9 for number of nodes and time.Hour for duration?

Reply:
Maybe we can have 3 states of thresholds and adjust thresholdBasedEstimationLimiter for them:
- negative: cannot add any nodes if at least one threshold is negative
- zero: can add as many nodes as you like, thresholds with return value 0 are ignored for current estimation
- positive: add no more than this number of nodes, it is the smallest possible value from all positive thresholds

What do you think?

Reply:
I'm not a fan of encoding such things into numerical values. But also I see that the "0 means unlimited" concept is pretty well established here.

I think this should work as long as we make sure that nodegroups that are not able to scale-up even a single node are not passed to the estimator. This is a pretty strong assumption though, I would add a comment about this in the Estimator call site.

Reply:
Now thresholds may return negative value and its is handled by thresholdBasedEstimationLimiter - f3dfeeeb47a021ca4b854be5363e1a1663313b3b